### ISSUE TITLE
Recursive polymorphic types can cause compiler to run out of memory

### ISSUE THREAD

**Issue by lukebemish** [flix/flix#6850]
Constructing a type such as the following:
```flix
enum ProblematicRecursive[a] {
    case Simple
    case Recurse(ProblematicRecursive[List[a]])
}
```
Will cause the compiler to run out of memory while compiling anything using it; presumably, this is due to it trying to resolve, say, `ProblematicRecursive[Int32]`, and then needing to resolve every various `ProblematicRecursive[List^n[Int32]]` to do so due to the `Recurse` case; I'm afraid I don't quite understand the flix compiler enough to have any more insight on this, but I figured I'd open this for posterity

**Comment by magnus-madsen** [comment]
Hi @lukebemish 

Thanks for reporting this. Looks like a stack overflow:

```
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.unification.TypeNormalization$.normalizeType(TypeNormalization.scala:48)  
        at ca.uwaterloo.flix.language.phase.MonoTypes$.$anonfun$specializeEnum$2(MonoTypes.scala:445)
        at scala.collection.immutable.List.map(List.scala:246)
        at ca.uwaterloo.flix.language.phase.MonoTypes$.specializeEnum(MonoTypes.scala:445)
        at ca.uwaterloo.flix.language.phase.MonoTypes$.visitInner$1(MonoTypes.scala:398)
        at ca.uwaterloo.flix.language.phase.MonoTypes$.visitType(MonoTypes.scala:411)
        at ca.uwaterloo.flix.language.phase.MonoTypes$.specializeCase(MonoTypes.scala:471)
        at ca.uwaterloo.flix.language.phase.MonoTypes$.$anonfun$specializeEnum$5(MonoTypes.scala:458)
        at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:676)
        at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:670)
        at scala.collection.AbstractIterable.foldLeft(Iterable.scala:933)
```

**Comment by magnus-madsen** [comment]
@JonathanStarup Could you take a look, with some haste? :) Thanks!

**Comment by magnus-madsen** [comment]
Minimal viable example:

```
enum ProblematicRecursive[a] {
    case Simple
    case Recurse(ProblematicRecursive[List[a]])
}

def main(): Unit \ IO = 
    let _ = ProblematicRecursive.Simple;
    println("Hello World!")

    
```

**Comment by JonathanStarup** [comment]
@magnus-madsen this is not a code bug, if you want to monomorphize `ProblematicRecursive` it will produce infinite enums. I'm discussing ideas with @mlutze but I'm not sure what to do in this case

**Comment by magnus-madsen** [comment]
I think we should monomorphize to erased types.

Assoc types have been dealt with in the previous phase, so it should be OK, no?

**Comment by JonathanStarup** [comment]
I think that only works if you ignore associated types. lets say that we should monomorphize `E[Magic[Int32]]`. Some case might use associated types to refer to `Int32`, in which case we can't just use `Object` in place of `Magic[Int32]`

**Comment by JonathanStarup** [comment]
Hmm I think that works yes. You prefer that over lazy case-monomorphization?

**Comment by JonathanStarup** [comment]
I think its easy to do cases lazily but still keep the enum. The cases are dependent on the generated interface but the interface doesn't care if all the cases are there

**Comment by JonathanStarup** [comment]
Fixed on master, tests added here #7182
